### DISCUSSION TITLE
Item Effect When Worn (MC Forge 1.20.1)

### DISCUSSION THREAD

**Discussion by Marquito3388** [TheIllusiveC4/Curios#366] [Q&A]
Hi! I am currently making a mod that requires Curios and I am having some trouble with my curio item. So I made it so I can equip the item BUT it wont give me the effect. I looked over some other mods and I decided to use the method that Botania did (which for me seemed simple compared to others). This is how my item class looks like right now:

```Java
package net.marquito.caa.item;

import net.minecraft.world.effect.MobEffectInstance;
import net.minecraft.world.effect.MobEffects;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import top.theillusivec4.curios.api.SlotContext;
import top.theillusivec4.curios.api.type.capability.ICurioItem;

public class DarkMatter extends Item implements ICurioItem {


    public DarkMatter(Properties pProperties) {
        super(pProperties);
    }

    public void onUnequipped(ItemStack stack, LivingEntity living) {
        MobEffectInstance effect = living.getEffect(MobEffects.FIRE_RESISTANCE);
        if (effect != null && effect.getAmplifier() == 0) {
            living.removeEffect(MobEffects.FIRE_RESISTANCE);
        }
    }

    public void curioTick(SlotContext slotContext, int index, LivingEntity entity, ItemStack stack) {

        if(entity instanceof Player player && !player.level().isClientSide){


            player.getActiveEffects();
            boolean hasPlayerFireResistance =
                    true;

            if (!hasPlayerFireResistance) {
                onUnequipped(stack, player);
            } else {
                if (player.getEffect(MobEffects.FIRE_RESISTANCE) != null) {
                    player.removeEffect(MobEffects.FIRE_RESISTANCE);
                }

                player.addEffect(new MobEffectInstance(MobEffects.FIRE_RESISTANCE, Integer.MAX_VALUE, 0, true, true));
            }

        }


        ICurioItem.super.curioTick(slotContext, stack);
    }
}
```

The item class makes sense to me but the item wont give me the Fire Resistance effect when I equip it. Is there something that I am missing? I saw that the documentation also says to add this to the mod main class inside the setup method:

```CuriosApi.registerCurio(CurioModItems.MY_ITEM, new CurioItem());```

 but mine is not working because it says that it expects an argument but I dont know what argument to pass. How do I do this?

**Comment by TheIllusiveC4** [comment]
```java
public void curioTick(SlotContext slotContext, int index, LivingEntity entity, ItemStack stack)
```

This is an incorrect method signature, it doesn't override anything from the `ICurioItem` interface so it doesn't run.  This is the correct method signature:
https://github.com/TheIllusiveC4/Curios/blob/2b34ad895c9ffc511eae2368b96d7538d7cfc58b/src/main/java/top/theillusivec4/curios/api/type/capability/ICurioItem.java#L77

> I saw that the documentation also says to add this to the mod main class inside the setup method:
> 
> CuriosApi.registerCurio(CurioModItems.MY_ITEM, new CurioItem());
> 
> but mine is not working because it says that it expects an argument but I dont know what argument to pass. How do I do this?

That step is unnecessary if you are implementing the `ICurioItem` interface directly on your item as you are.

**Reply by Marquito3388** [reply]
I was looking over Botania documentation and they use this method (the one I am using) for their Invisibility Cloak so I thought it would also make sense to use. I am just not sure how I would write the curioTick method to give the player Fire Resistance while having the item equipped. And also, what would go in the onEquip and onUnequip. The curioTick method would just act as a "while" loop then? I dont know how the could would look like.

```java default void curioTick(SlotContext slotContext, ItemStack stack) {
        curioTick(slotContext.identifier(), slotContext.index(), slotContext.entity(), stack);
    }
    
 default void onEquip(SlotContext slotContext, ItemStack prevStack, ItemStack stack) {
        onEquip(slotContext.identifier(), slotContext.index(), slotContext.entity(), stack);
    }

default void onUnequip(SlotContext slotContext, ItemStack newStack, ItemStack stack) {
        onUnequip(slotContext.identifier(), slotContext.index(), slotContext.entity(), stack);
    }

**Reply by Marquito3388** [reply]
Basically, I am wondering how I can convert my code into this curioTick. Looking over tutorials (I found an old one from 1.16.5), they check if the player is server side or not. But apparently this method you showed here already considers that.

**Reply by TheIllusiveC4** [reply]
You've already written the method, your main issue is that you have the incorrect method signature as I mentioned above. You are using `curioTick(SlotContext slotContext, int index, LivingEntity entity, ItemStack stack)`, which is not a method inherited from the `ICurioItem` interface. You need to look at the `ICurioItem` interface to see the correct methods:
https://github.com/TheIllusiveC4/Curios/blob/1.20.x/src/main/java/top/theillusivec4/curios/api/type/capability/ICurioItem.java

I quoted the correct method signature in my original reply, all you need to do is change your method signature to match it so that it gets properly overridden.

**Reply by Marquito3388** [reply]
so you mean something like this?


```java public void curioTick(SlotContext slotContext, ItemStack stack) {


        if(slotContext.entity() instanceof Player player && !player.level().isClientSide){


            player.getActiveEffects();
            boolean hasPlayerFireResistance =
                    true;

            if (!hasPlayerFireResistance) {
                onUnequipped(stack, player);
            } else {
                if (player.getEffect(MobEffects.FIRE_RESISTANCE) != null) {
                    player.removeEffect(MobEffects.FIRE_RESISTANCE);
                }

                player.addEffect(new MobEffectInstance(MobEffects.FIRE_RESISTANCE, Integer.MAX_VALUE, 0, true, true));
            }

        }


        ICurioItem.super.curioTick(slotContext, stack);
    }

**Reply by TheIllusiveC4** [reply]
No, that's the body of your method. I'm talking about your method signature. 

```
public void curioTick(SlotContext slotContext, int index, LivingEntity entity, ItemStack stack)
```

That's what you have now. It should be:
```
public void curioTick(SlotContext slotContext, ItemStack stack)
```

**Reply by Marquito3388** [reply]
Sorry I forgot to include the method as well. But I did exactly what you mentioned and it didnt work. Ill paste it here for you to see. But is the body wrong though? It is still not working. Do I still have to include the `CuriosApi.registerCurio(CurioModItems.MY_ITEM, new CurioItem());` ?


 ```java public class DarkMatter extends Item implements ICurioItem {


    public DarkMatter(Properties pProperties) {
        super(pProperties);
    }

    public void onUnequipped(ItemStack stack, LivingEntity living) {
        MobEffectInstance effect = living.getEffect(MobEffects.FIRE_RESISTANCE);
        if (effect != null && effect.getAmplifier() == 0) {
            living.removeEffect(MobEffects.FIRE_RESISTANCE);
        }
    }

    public void curioTick(SlotContext slotContext, ItemStack stack) {


        if(slotContext.entity() instanceof Player player && !player.level().isClientSide){


            player.getActiveEffects();
            boolean hasPlayerFireResistance =
                    true;

            if (!hasPlayerFireResistance) {
                onUnequipped(stack, player);
            } else {
                if (player.getEffect(MobEffects.FIRE_RESISTANCE) != null) {
                    player.removeEffect(MobEffects.FIRE_RESISTANCE);
                }

                player.addEffect(new MobEffectInstance(MobEffects.FIRE_RESISTANCE, Integer.MAX_VALUE, 0, true, true));
            }

        }
    }
}`

**Reply by TheIllusiveC4** [reply]
You will need to setup breakpoints or debug lines inside your method body to see if it's running the code at all. That way you can figure out if it's a problem with the setup or a problem with the method body.
